javatests/arcs/core/data/expression/TypeCheckerTest.kt

```diff
@@ -0,0 +1,300 @@
+class TypeCheckerTest {
+  /**
+   * Checks that the resulting type of the expression is compatible with the expected type. For
+   * example, a byte is assignable to an int, but an int is not assignable to a byte. A union
+   * of A|B is assignable to A|B|C but not vice-versa.
+   */
+  fun checkAssignableTo(
+    expr: Expression<*>,
+    expectedType: InferredType,
+    scope: Expression.Scope = MapScope<InferredType>("root", mapOf()),
+    params: Expression.Scope = ParameterScope()
+  ) {
+    val inferredType = expr.accept(TypeEvaluator(params), scope)
+    assertTrue(expectedType.isAssignableFrom(inferredType), "$expectedType != $inferredType")
+  }
+
+  @Test
+  fun check_literals() {
+    checkAssignableTo(1.0.asExpr(), DoubleType)
+    checkAssignableTo(1.0.asExpr(), NumberType)
+
+    checkAssignableTo(1.0f.asExpr(), FloatType)
+    checkAssignableTo(1.0f.asExpr(), DoubleType)
+    checkAssignableTo(1.0f.asExpr(), NumberType)
+
+    checkAssignableTo(BigInt("1").asExpr(), BigIntType)
+    checkAssignableTo(BigInt("1").asExpr(), FloatType)
+    checkAssignableTo(BigInt("1").asExpr(), DoubleType)
+    checkAssignableTo(BigInt("1").asExpr(), NumberType)
+
+    checkAssignableTo(1L.asExpr(), LongType)
+    checkAssignableTo(1L.asExpr(), BigIntType)
+    checkAssignableTo(1L.asExpr(), DoubleType)
+    checkAssignableTo(1L.asExpr(), FloatType)
+    checkAssignableTo(1L.asExpr(), NumberType)
+
+    checkAssignableTo(1.asExpr(), IntType)
+    checkAssignableTo(1.asExpr(), BigIntType)
+    checkAssignableTo(1.asExpr(), LongType)
+    checkAssignableTo(1.asExpr(), DoubleType)
+    checkAssignableTo(1.asExpr(), FloatType)
+    checkAssignableTo(1.asExpr(), NumberType)
+
+    checkAssignableTo(1.toShort().asExpr(), ShortType)
+    checkAssignableTo(1.toShort().asExpr(), IntType)
+    checkAssignableTo(1.toShort().asExpr(), BigIntType)
+    checkAssignableTo(1.toShort().asExpr(), LongType)
+    checkAssignableTo(1.toShort().asExpr(), FloatType)
+    checkAssignableTo(1.toShort().asExpr(), DoubleType)
+    checkAssignableTo(1.toShort().asExpr(), NumberType)
+
+    checkAssignableTo(1.toByte().asExpr(), ByteType)
+    checkAssignableTo(1.toByte().asExpr(), ShortType)
+    checkAssignableTo(1.toByte().asExpr(), IntType)
+    checkAssignableTo(1.toByte().asExpr(), BigIntType)
+    checkAssignableTo(1.toByte().asExpr(), LongType)
+    checkAssignableTo(1.toByte().asExpr(), FloatType)
+    checkAssignableTo(1.toByte().asExpr(), DoubleType)
+    checkAssignableTo(1.toByte().asExpr(), NumberType)
+    checkAssignableTo("foo".asExpr(), TextType)
+    checkAssignableTo(true.asExpr(), BooleanType)
+  }
+
+  @Test
+  fun check_unary_ops() {
+    checkAssignableTo(-(1.asExpr()), IntType)
+    checkAssignableTo(-(1.0.asExpr()), DoubleType)
+    checkAssignableTo(!(true.asExpr()), BooleanType)
+  }
+
+  @Test
+  fun check_binary_ops() {
+    checkAssignableTo(1.asExpr() + 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() + 2.asExpr(), DoubleType)
```

Comment:
If there was a bug in widening and `1.0.asExpr() + 2.asExpr()` got `IntType` inferred, this assertion would pass. This makes me wonder why you perform all the tests in terms of `checkAssignableTo`, why don't do simple equality here?

Reply:
Lazyiness, I didn't get around to writing a helper function to check equality yet.